Fix Telegram Bot send file to multiple targets, snapshots of HA cameras, variable templating, digest auth

azogue opened the pull request:
## Description:

Fix some problems introduced with the changes in `telegram_bot` and `notify.telegram` of 0.45 and 0.45.1, and some enhancements:

- Allow templating in `url`, `file`, `caption`, `longitude`, `latitude` fields.
- Fix sending HA camera snapshots by retrying the HA API request (it returns some 500's before working for some cameras in Home Assistant, not all)
- Fix send file to multiple locations
- Easier inline keyboards from yaml files.
- HttpDigest authentication for url’s

**Related issues:** 
- fix answering to callback queries (KeyError)
- fixes #7703 Location information not being sent to Telegram
- fixes #7413 Cannot send specific jpg via telegram-bot
- fixes #7395 Telegram does not send photo if URL requires HttpDigestAuth

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#2714

## Notification Examples for `configuration.yaml`:
```yaml
telegram_bot:
  platform: polling
  api_key: !secret telegram_api
  allowed_chat_ids:
    - !secret telegram_bot_chat_id_admin
    - !secret telegram_bot_group

notify:
  - platform: telegram
    name: telegram_bot
    chat_id: !secret telegram_bot_chat_id_admin

script:
  # Telegram Bot component
  test_telegram_bot_location:
    alias: Test Telegram Bot multi-target location template
    sequence:
      - service: telegram_bot.send_location
        data:
          caption: 'Test location'
          target:
            - !secret telegram_bot_chat_id_admin
            - !secret telegram_bot_group
          latitude: '{{ states.device_tracker.iphone.attributes.latitude|float }}'
          longitude: '{{ states.device_tracker.iphone.attributes.longitude|float }}'
          disable_notification: true

  test_telegram_bot_document:
    alias: Test Telegram Bot multi-target document
    sequence:
      - service: telegram_bot.send_document
        data:
          caption: 'Test document'
          target:
            - !secret telegram_bot_chat_id_admin
            - !secret telegram_bot_group
          file: '/home/homeassistant/.homeassistant/home-assistant.log'
          disable_notification: true

  test_telegram_bot_image:
    alias: Test Telegram Bot multi-target image
    sequence:
      - service: telegram_bot.send_photo
        data:
          target:
            - !secret telegram_bot_chat_id_admin
            - !secret telegram_bot_group
          caption: 'Test image'
          url: 'http://localhost:8123{{ states.camera.ha_camera.attributes.entity_picture}}'
          disable_notification: true

  # Notify component
  test_notify_telegram_location:
    alias: Test notify.telegram location template
    sequence:
      - service: notify.telegram_bot
        data:
          message: "Not used here but needed..."
          data:
            location:
              latitude: '{{ states.device_tracker.iphone.attributes.latitude|float }}'
              longitude: '{{ states.device_tracker.iphone.attributes.longitude|float }}'

  test_notify_telegram_image:
    alias: Test notify.telegram multiple image
    sequence:
      - service: notify.telegram_bot
        data:
          message: "Not used here but needed..."
          data:
            photo:
              - url: 'https://upload.wikimedia.org/wikipedia/en/5/58/Penny_test.jpg'
                caption: 'Test img 1. C1'
              - url: 'http://localhost:8123{{ states.camera.ha_camera.attributes.entity_picture}}'
                caption: 'Test img 1. C2'
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)
  - [X] Tests have been done manually to verify that the new code works. (There aren't tests for the Telegram components)

azogue committed:
fix double template rendering when messages come from notify.telegram

azogue committed:
fix 'chat' information not present in callback queries

azogue committed:
better inline keyboards with yaml

To make a row of InlineKeyboardButtons you pass:
- a list of tuples like: `[(text_b1, data_callback_b1), (text_b2, data_callback_b2), ...]
- a string like: `/cmd1, /cmd2, /cmd3`
- or a string like: `text_b1:/cmd1, text_b2:/cmd2`

Example:
```yaml
data:
   message: 'TV is off'
   disable_notification: true
   inline_keyboard:
     - TV ON:/service_call switch.turn_on switch.tv, Other:/othercmd
     - /help, /init
```

azogue committed:
fix send file to multiple targets

azogue committed:
fix message templating, multiple file targets, HA cameras

- Allow templating for caption, url, file, longitude and latitude fields
- Fix send a file to multiple targets
- Load data with some retrying for HA cameras, which return 500 one or two times sometimes (generic cams, always!).
- Doc in services for new inline keyboards yaml syntax: `Text button:/command`

azogue committed:
HttpDigest authentication as proposed in #7396

MartinHjelmare commented on homeassistant/components/telegram_bot/__init__.py:
```diff
@@ -296,48 +333,58 @@ def _get_msg_ids(self, msg_data, chat_id):
+            try:
+                chat_ids = [int(t) for t in target
+                            if int(t) in self.allowed_chat_ids]
+                if len(chat_ids) > 0:
```
```py
if chat_ids:
```

MartinHjelmare commented on homeassistant/components/telegram_bot/__init__.py:
```diff
@@ -296,48 +333,58 @@ def _get_msg_ids(self, msg_data, chat_id):
+                if len(chat_ids) > 0:
+                    return chat_ids
+                _LOGGER.warning("Unallowed targets: %s", target)
+            except (ValueError, TypeError):
```
When could this happen? Both target and allowed chat ids have been validated as list of integers before this.

MartinHjelmare commented on homeassistant/components/telegram_bot/__init__.py:
```diff
@@ -296,48 +333,58 @@ def _get_msg_ids(self, msg_data, chat_id):
+                    return chat_ids
+                _LOGGER.warning("Unallowed targets: %s", target)
+            except (ValueError, TypeError):
+                _LOGGER.warning("Bad target data: %s, using default: %s",
```
Shouldn't the warning about default  be included in the above warning instead? The warning here should never happen.

MartinHjelmare commented on homeassistant/components/telegram_bot/__init__.py:
```diff
@@ -446,20 +493,26 @@ def answer_callback_query(self, message, callback_query_id,
+        params = self._get_msg_kwargs(kwargs)
+        caption = kwargs.get(ATTR_CAPTION)
+        func_send = self.bot.sendPhoto if is_photo else self.bot.sendDocument
         file = load_data(
```
Pick another variable name than file. It's reserved in python.

MartinHjelmare replied:
This one is still here. :wink:

azogue replied:
fixed!  😓

MartinHjelmare commented on homeassistant/components/telegram_bot/__init__.py:
```diff
@@ -152,24 +157,41 @@
 
 
-def load_data(url=None, file=None, username=None, password=None):
+def load_data(url=None, file=None, username=None, password=None,
```
Pick another parameter name than `file`. It's reserved in python.

MartinHjelmare commented on homeassistant/components/telegram_bot/__init__.py:
```diff
@@ -152,24 +157,41 @@
+                        params["auth"] = HTTPDigestAuth(username, password)
+                    else:
+                        params["auth"] = HTTPBasicAuth(username, password)
+                req = requests.get(url, **params)
```
Now there's no request without authentication. Is that intended?

azogue replied:
authentication params are included only when are present:
```python
params = {"timeout": 15}
if username is not None and password is not None:
    if authentication == HTTP_DIGEST_AUTHENTICATION:
        params["auth"] = HTTPDigestAuth(username, password)
    else:
        params["auth"] = HTTPBasicAuth(username, password)
```
What do you mean?

MartinHjelmare replied:
You're correct. I missed that point, sorry.

azogue commented on homeassistant/components/telegram_bot/__init__.py:
```diff
@@ -296,48 +333,58 @@ def _get_msg_ids(self, msg_data, chat_id):
-                _LOGGER.warning("BAD TARGET %s, using default: %s",
+            try:
+                chat_ids = [int(t) for t in target
+                            if int(t) in self.allowed_chat_ids]
```
It's true, this is old, I'm going to simplify it

azogue committed:
review changes

- Don't use `file` as variable name.
- For loop
- Simplify filter allowed `chat_id`s.

azogue committed:
Don't use `file` as variable name!

azogue committed:
make params outside the while loop

MartinHjelmare reviewed (approved):
Nice!

azogue committed:
fix chat_id validation when editing sent messages

azogue commented:
@MartinHjelmare, sorry, I found one more when testing the examples in the doc.

UltraSub commented:
Nice! Is there any chance this will go into a 45.2 release?

MartinHjelmare commented:
@azogue Ready to merge?

azogue commented:
Yes!
